src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
+        content[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int listSize = list.size();
+        int currentLength = content.length;
+        if (listSize > currentLength - size) {
+            int missingCells = Math.max(listSize - (currentLength - size), currentLength >> 1);
+            grow(missingCells);
+        }
 
+        T[] listArray = convertListToArray(list);
+        System.arraycopy(listArray, 0, content, size, listSize);
+        size += listSize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isInRange(index)) {
+            return content[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("The index is out of bounds");
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!isInRange(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Can't set value, because the index "
+                    + "is out of bounds");
+        }
+        content[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isInRange(index)) {
+            T deletedValue = content[index];
+            System.arraycopy(content, index + 1, content, index, size-- - (index + 1));
+            return deletedValue;
+        }
+        throw new ArrayListIndexOutOfBoundsException("The index is out of list's range");
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getValueIndex(element);
+        if (index == NOT_FOUND_INDEX) {
+            throw new NoSuchElementException("There is no such element in array.");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        increaseContentLengthByNumber(content.length >> 1);
+    }
+
+    private void grow(int num) {
+        increaseContentLengthByNumber(num);
+    }
+
+    private void increaseContentLengthByNumber(int extraCells) {
+        int oldLength = content.length;
+        int newLength = oldLength + extraCells;
+        T[] tempArray = (T[]) new Object[newLength];
+        System.arraycopy(content, 0, tempArray, 0, oldLength);
+        content = tempArray;
+    }
+
+    private T[] convertListToArray(List<T> list) {
+        int listSize = list.size();
+        T[] array = (T[]) new Object[listSize];
+        for (int i = 0; i < listSize; i++) {
+            array[i] = list.get(i);
+        }
+        return array;
+    }
+
+    private boolean isInRange(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private int getValueIndex(T value) {
+        for (int i = 0; i < content.length; i++) {
+            if (Objects.equals(content[i], value)) {
```

Comment:
We're not allowed using Objects.equals() method in this task. Create your own implementation

Reply:
Done.